src/os/bluestore/BlueStore.cc

```diff
@@ -8287,56 +8287,189 @@ int BlueStore::repair_omap_upgrade()
   }
   if (r < 0)
     goto out_scan;
-
   {
+    // we assume that all entries under PREFIX_PERPG_OMAP are valid, that is
+    // each starts with: u64 pool_id + u32 hash + u64 nid
     auto it = db->get_iterator(PREFIX_PERPG_OMAP);
     if (it) {
+      uint64_t prev_pool;
+      uint64_t prev_hash;
+      uint64_t prev_nid;
+      bool expect_next_object = true;
+      bool expect_header_copy = false;
       size_t cnt = 0;
+      size_t problems = 0;
       KeyValueDB::Transaction txn = db->get_transaction();
       it->lower_bound(std::string());
       while(it->valid()) {
         auto key = it->key();
         auto val = it->value();
-        it->next();
-        auto p1 = key.find_first_of('.');
-        if (p1 == string::npos || p1 <= 8) {
-          continue;
-        }
-        auto p2 = key.find_first_of('-', p1 + 1);
-        if (p2 == string::npos) {
-          p2 = key.find_first_of('.', p1 + 1);
-        }
-        if (p2 == string::npos || p1 + 8 > p2) {
-          continue;
-        }
-        string s1 = key.substr(p1 - 8, 8);
-        string s2 = key.substr(p2 - 8, 8);
-        if (s1 != s2) {
-          continue;
-        }
-        txn->rmkey(PREFIX_PERPG_OMAP, key);
-        if (key[p2] == '.') {
-          auto new_key = key;
-          new_key.erase(p1, p2 - p1);
-          dout(0) << __func__ << " " << pretty_binary_string(key)
-                  << " -> " << pretty_binary_string(new_key) << dendl;
-          txn->set(PREFIX_PERPG_OMAP, new_key, val);
-        } else {
-          dout(0) << __func__ << " " << pretty_binary_string(key)
-                  << " rm " << dendl;
-        }
-        cnt++;
-        if (cnt >= 1024) {
-          key = it->key();
-          db->submit_transaction_sync(txn);
-          it->lower_bound(key);
-          txn = db->get_transaction();
-          cnt = 0;
-        }
+	dout(20) << __func__ << " key=" << pretty_binary_string(key) << dendl;
+	it->next();
+	ceph_assert(key.length() >= 8 + 4 + 8 + 1);
```

Comment:
let's just report invalid key and continue here....

Reply:
right....